thorlcr/slave/traceslave.hpp

```diff
@@ -0,0 +1,226 @@
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#ifndef TRACESLAVE_HPP
+#define TRACESLAVE_HPP
+
+#include <limits.h>
+#include "thmem.hpp"
+
+class CTracingThorDataLink : implements CInterfaceOf<IThorDataLink>
+{
+private:
+    class CThorRowHistory : public OwnedConstThorRow
+    {
+        unsigned sequence;
+        unsigned msTickStamp;
+
+    public:
+        inline CThorRowHistory() : sequence(0), msTickStamp(0) { }
+
+        inline void set(const void * _row, unsigned _sequence, unsigned _mstick)
+        {
+            sequence = _sequence;
+            msTickStamp = _mstick;
+            OwnedConstThorRow::set(_row);
+        }
+        inline unsigned getSequence() const { return sequence; }
+        inline unsigned getmsTickStamp() const { return msTickStamp; }
+        inline void getAgeInMs(StringBuffer &ageAsAString) const
+        {
+            const unsigned _msTick = msTick();
+            const unsigned age = (_msTick>=msTickStamp)?(_msTick-msTickStamp):(UINT_MAX-_msTick)+msTickStamp;
+            const unsigned hrs = age/(60*60*1000);
+            const unsigned mins = (age/(60*1000))-hrs;
+            const double secs = (double)(age-hrs-mins)/1000;
+
+            ageAsAString.clear();
+            if (hrs) ageAsAString.appendf("%uh ", hrs);
+            if (mins||hrs) ageAsAString.appendf("%um ", mins);
+            ageAsAString.appendf("%2.3fs", secs);
+        }
+    };
+    class TraceQueue
+    {
+    public:
+        CThorRowHistory *rowBuffer;
+        unsigned rowBufferSize;
+        unsigned rowTailp;
+
+        TraceQueue() : rowBuffer(NULL), rowBufferSize(0), rowTailp(0){ }
+
+        ~TraceQueue()
+        {
+            delete [] rowBuffer;
+        }
+        void init(unsigned _bufsize)
+        {
+            if (!rowBuffer && _bufsize)
+            {
+                rowBufferSize = _bufsize;
+                rowBuffer = new CThorRowHistory[rowBufferSize];
+                rowTailp = 0;
+            }
+        }
+        inline void enqueue(const void *row, unsigned sequence, unsigned _msTick)
+        {
+            // NOTE - updates then sets rowHeadp (atomically) so that the code
+            // below (ignoring oldest record) is safe
+            rowBuffer[rowTailp].set(row,sequence,_msTick);
+            unsigned nextTailp = rowTailp + 1;
+            if (nextTailp==rowBufferSize)
+                nextTailp = 0;
+            rowTailp = nextTailp;
+        }
+        inline void enqueue(const CThorRowHistory &_rowBuffer)
+        {
+            enqueue(_rowBuffer.get(),_rowBuffer.getSequence(),_rowBuffer.getmsTickStamp());
```

Comment:
Still concerned extra complication and usefulness  of sequence and also now timestamp (the cost of it)